Added calls to ZooZap when shutting down groups of servers

dlmarion opened the pull request:
Closes #5178

keith-turner commented on assemble/bin/accumulo-cluster:
```diff
@@ -471,6 +471,12 @@ function control_services() {
       done
+      if [[ $operation == "stop" || $operation == "kill" ]]; then
+        echo "Cleaning tablet server entries from zookeeper for resource group $group"
+        if ! isDebug; then
+          "$accumulo_cmd" org.apache.accumulo.server.util.ZooZap -tservers -group "$group"
+        fi
```
could `debugOrRun` be used here? Thinking it could print the command that would be run in debug mode.

dlmarion replied:
We could do that. I surveyed the other places in `accumulo-cluster` and we are not using `debugOrRun` in all cases where we are executing a command.

dlmarion replied:
I applied this change in 18844f8.

ddanielr replied:
`./accumulo-cluster stop --tservers=extra --local --dry-run`
returned: 
```
bin/accumulo' 'org.apache.accumulo.server.util.ZooZap' '-verbose' '-tservers' '-group' 'extra'
```

dlmarion replied:
@ddanielr - are you suggesting there is an issue? or were you just answering @keith-turner 's question about what it would look like?

ddanielr replied:
I'm just answering @keith-turner 's question. The debug option seems to work well!

keith-turner reviewed (approved):
Looks good, did you manually test this?  Wondering what the output would look like and if there would be an indication that zoozap ran.  That lead to the comment about debugOrRun

dlmarion commented:
> Looks good, did you manually test this? Wondering what the output would look like and if there would be an indication that zoozap ran. That lead to the comment about debugOrRun

I applied your suggestion and made some changes to ZooZap in 18844f8 to clean up zookeeper in a consistent manner for the different server types. ZooZap was doing one thing for compactors, and something else for tservers and sservers. I changed ZooZap so that in all cases it does a recursive delete on the resource group.

Below is the output of `accumulo-cluster stop` on my local node, with the log4j lines removed.

```
Stopping Accumulo cluster...
Accumulo shut down cleanly
Utilities and unresponsive servers will shut down in 5 seconds (Ctrl-C to abort)
Executing stop on tablet servers for group default ....Cleaning tablet server entries from zookeeper for resource group default
Stopping service process: tserver_default_1
Stopping tserver_default_1 on 1.2.3.4
done
Executing stop on managers
Executing stop on garbage collectors
Executing stop on monitors
Executing stop on scan servers for group default
Stopping service process: manager_default_1
Stopping manager_default_1 on 1.2.3.4
Stopping service process: gc_default_1
Stopping gc_default_1 on 1.2.3.4
Cleaning scan server entries from zookeeper for resource group default
Stopping service process: monitor_default_1
Stopping monitor_default_1 on 1.2.3.4
Stopping service process: sserver_default_1
Stopping sserver_default_1 on 1.2.3.4
Executing stop on compactors for group default
Cleaning compactor entries from zookeeper for resource group default
Stopping service process: compactor_default_1
Stopping compactor_default_1 on 1.2.3.4
Deleting compactor /accumulo/063fa7a3-0eda-409e-89b8-6d559d6269cf/compactors/default from zookeeper
```
I think it shows several problems where the ZooKeeper cleanup is inconsistent.

  1. In the case of the compactor and sserver, you see a message "Deleting ..... from zookeeper". This is coming from ZooZap, but you don't see it for the tserver. Looking at the tserver log it shows that a stop was requested, performed a graceful shutdown, and removed it's own lock from ZooKeeper. When ZooZap was executed the tserver address I think was empty, so it did not remove the tserver resource group default path. I think this may have been cleaned up by the `admin stopAll` code path, but not entirely sure at the moment.
  2. For the compactor case you see that the  you see the following:
  ```
  Executing stop on compactors for group default
  Cleaning compactor entries from zookeeper for resource group default
  Stopping service process: compactor_default_1
  Stopping compactor_default_1 on 1.2.3.4
  Deleting compactor /accumulo/063fa7a3-0eda-409e-89b8-6d559d6269cf/compactors/default from zookeeper
  ```
  The two lines about remove entries from ZooKeeper are from the ZooZap command which comes after the code in `accumulo-cluster` that performs the stop on the compactor process. In my setup I'm using a remote address for the processes, so an ssh command is being used to run the command to stop the compactor. I think it's possible that the ZooKeeper paths could be removed before the processes are stopped in this case.

keith-turner commented:
> I think it's possible that the ZooKeeper paths could be removed before the processes are stopped in this case.

Any insight into why this is happening?  Looking at the accumulo-cluster script it seems like the call to `accumulo-service` happens before the call to zoozap.

dlmarion commented:
> > I think it's possible that the ZooKeeper paths could be removed before the processes are stopped in this case.
> 
> Any insight into why this is happening? Looking at the accumulo-cluster script it seems like the call to `accumulo-service` happens before the call to zoozap.

I believe that it's the recent changes to the `accumulo-cluster` script with the `ssh` command. IIRC it uses the options `qnf`, which if my memory serves me, puts the command into the background in another process. I believe we did this to help speed up processing of commands on systems with a large number of server processes, but it also means that the `accumulo-cluster` script will continue with the commands not having completed. Based on this, I'm thinking we may not want to make this change, or I need to look at doing it differently.

dlmarion commented:
e5d7224 introduces a mechanism that will wait for all of the ssh commands to finish before calling ZooZap. The `shellcheck` GitHub action is failing with the following:
```
^-- SC2009 (info): Consider using pgrep instead of grepping ps output.
```

I'm thinking I should just suppress this. I'm not sure if `pgrep` is installed by default on all operating systems.

ddanielr commented on assemble/bin/accumulo-cluster:
```diff
@@ -520,6 +533,13 @@ function control_services() {
+        # to complete before zapping the nodes in ZooKeeper
+        ssh_wait
+        echo "Cleaning scan server entries from zookeeper for resource group $group"
+        debugOrRun "$accumulo_cmd" org.apache.accumulo.server.util.ZooZap -verbose -sservers -group "$group"
```
same `-local` condition handling comment

ddanielr commented on assemble/bin/accumulo-cluster:
```diff
@@ -538,14 +558,19 @@ function control_services() {
+        # to complete before zapping the nodes in ZooKeeper
+        ssh_wait
+        echo "Cleaning compactor entries from zookeeper for resource group $group"
+        debugOrRun "$accumulo_cmd" org.apache.accumulo.server.util.ZooZap -verbose -compactors -group "$group"
```
same `-local` condition handling comment

ddanielr commented on server/base/src/main/java/org/apache/accumulo/server/util/ZooZap.java:
```diff
@@ -136,21 +135,12 @@ public void zap(SiteConfiguration siteConf, String... args) {
           try {
             Set<ServiceLockPath> tserverLockPaths =
                 context.getServerPaths().getTabletServer(rgp, AddressSelector.all(), false);
```
We could do something like the following to support a host filter:
```
AddressSelector hostLimiter;
if (!opts.hostname.isEmpty()) {
   hostLimiter = AddressSelector.matching(hp -> hp.startsWith(opts.hostname)); 
} else { 
   hostLimiter = AddressSelector.all();
}
   ```

dlmarion commented on assemble/bin/accumulo-cluster:
```diff
@@ -471,6 +477,13 @@ function control_services() {
+        # to complete before zapping the nodes in ZooKeeper
+        ssh_wait
+        echo "Cleaning tablet server entries from zookeeper for resource group $group"
+        debugOrRun "$accumulo_cmd" org.apache.accumulo.server.util.ZooZap -verbose -tservers -group "$group"
```
Really good catch! I think your suggestion of `could be fixed by adding logic to check for the --local arg and only removing entries if its a cluster-wide action` should be sufficient. The issue I'm trying to solve here, re-adding functionality introduced in #5114 and subsequently removed in the Great Script Refactor, did not address the local case.

ddanielr reviewed (approved):
Tested code functionality and it seems to work as expected.

The issue with deleting locks regardless of `--local` can be handled in a follow-on if we want to get these changes pulled in first.

dlmarion commented:
> The issue with deleting locks regardless of `--local` can be handled in a follow-on if we want to get these changes pulled in first.

I think it's a blocker, I'm fixing it now.

dlmarion commented:
Ok, I think I just need an opinion on my https://github.com/apache/accumulo/pull/5321#issuecomment-2668975289 above. I'm thinking of just suppressing the shellcheck suggestion.

ddanielr commented:
> [e5d7224](https://github.com/apache/accumulo/commit/e5d7224495bf610fd75f31c4377755c6350c73f5) introduces a mechanism that will wait for all of the ssh commands to finish before calling ZooZap. The `shellcheck` GitHub action is failing with the following:
> 
> ```
> ^-- SC2009 (info): Consider using pgrep instead of grepping ps output.
> ```
> 
> I'm thinking I should just suppress this. I'm not sure if `pgrep` is installed by default on all operating systems.

I can't find anything that definitively says that `pgrep` is installed by default.
There's references pointing to "pre-installed on most operating systems" but that's not absolute. 
I vote for suppressing it as you aren't doing anything complicated in the grep.

dlmarion commented:
> Tested new changes and when `--local` is passed no remote server locks are removed.

Thanks @ddanielr !

ctubbsii commented:
> [e5d7224](https://github.com/apache/accumulo/commit/e5d7224495bf610fd75f31c4377755c6350c73f5) introduces a mechanism that will wait for all of the ssh commands to finish before calling ZooZap. The `shellcheck` GitHub action is failing with the following:
> 
> ```
> ^-- SC2009 (info): Consider using pgrep instead of grepping ps output.
> ```
> 
> I'm thinking I should just suppress this. I'm not sure if `pgrep` is installed by default on all operating systems.

procps-ng, which ships pgrep and pkill, is a dependency of plymouth for graphical boot, so it can be presumed to be installed on workstations. It might not be installed on a minimal server install, but it's also a dependency of initscripts, so it can be presumed to be installed by default on Fedora/RHEL/CentOS and derivatives.

3.3.15 and earlier of procps-ng has a bug where it can't find commands whose length exceeds 4096 (which includes ZooKeeper). See https://bugzilla.redhat.com/show_bug.cgi?id=1782309
